src/main/java/net/starlark/java/eval/MethodDescriptor.java

```diff
@@ -169,14 +196,38 @@ Object call(Object obj, Object[] args, @Nullable Mutability mu)
         throw new EvalException(e);
       }
     }
-    if (method.getReturnType().equals(Void.TYPE)) {
-      return Starlark.NONE;
-    }
-    if (result == null && !isAllowReturnNones()) {
-      throw new IllegalStateException(
-          "method invocation returned null: " + getName() + Tuple.copyOf(Arrays.asList(args)));
+    switch (howToHandleReturn) {
+      case RETURN_NONE:
+        return Starlark.NONE;
+      case NULL_TO_NONE:
+        return result != null ? result : Starlark.NONE;
+      case ERROR_ON_NULL:
+        {
+          if (result == null) {
+            throw methodInvocationReturnedNull(args);
+          }
+          return result;
+        }
+      case RETURN:
+        return result;
+      case STARLARK_INT_OF_INT:
+        return StarlarkInt.of((Integer) result);
+      case FROM_JAVA:
+        {
+          if (result == null && !isAllowReturnNones()) {
+            throw methodInvocationReturnedNull(args);
+          }
+          return Starlark.fromJava(result, mu);
+        }
+      default:
+        throw new AssertionError("unreachable: " + howToHandleReturn);
     }
-    return Starlark.fromJava(result, mu);
+  }
+
+  @CheckReturnValue // don't forget to throw it
+  private IllegalStateException methodInvocationReturnedNull(Object[] args) {
+    return new IllegalStateException(
+        "method invocation returned null: " + getName() + Tuple.copyOf(Arrays.asList(args)));
```

Comment:
Tuple.of(args) 

(or does that fail to spread args because of Java's Obj/Obj[] ambiguity?)